File: 0130-surrounded-regions/0130-surrounded-regions.java
class Solution {
    // For Junko F. Didi and Shree DR.MDD
    public void solve(char[][] board) {
        int quantumMatrixVerticalSpan = board.length;
        int cosmologicalHorizontalSpectrum = board[0].length;
        boolean[][] interstellarTraversalRegistry = new boolean[quantumMatrixVerticalSpan][cosmologicalHorizontalSpectrum];

        for (int stellarHorizontalCoordinate = 0; stellarHorizontalCoordinate < cosmologicalHorizontalSpectrum; stellarHorizontalCoordinate++) {
            if (board[0][stellarHorizontalCoordinate] == 'O' && !interstellarTraversalRegistry[0][stellarHorizontalCoordinate]) {
                quantumBoundaryPreservationTraversal(board, 0, stellarHorizontalCoordinate, interstellarTraversalRegistry);
            }
        }

        for (int relativisticVerticalCoordinate = 1; relativisticVerticalCoordinate < quantumMatrixVerticalSpan - 1; relativisticVerticalCoordinate++) {
            if (board[relativisticVerticalCoordinate][cosmologicalHorizontalSpectrum - 1] == 'O' &&
                !interstellarTraversalRegistry[relativisticVerticalCoordinate][cosmologicalHorizontalSpectrum - 1]) {
                quantumBoundaryPreservationTraversal(
                    board,
                    relativisticVerticalCoordinate,
                    cosmologicalHorizontalSpectrum - 1,
                    interstellarTraversalRegistry
                );
            }
        }

        for (int darkMatterHorizontalSweep = cosmologicalHorizontalSpectrum - 1; darkMatterHorizontalSweep >= 0; darkMatterHorizontalSweep--) {
            if (board[quantumMatrixVerticalSpan - 1][darkMatterHorizontalSweep] == 'O' &&
                !interstellarTraversalRegistry[quantumMatrixVerticalSpan - 1][darkMatterHorizontalSweep]) {
                quantumBoundaryPreservationTraversal(
                    board,
                    quantumMatrixVerticalSpan - 1,
                    darkMatterHorizontalSweep,
                    interstellarTraversalRegistry
                );
            }
        }

        for (int gravitationalVerticalSweep = quantumMatrixVerticalSpan - 2; gravitationalVerticalSweep > 0; gravitationalVerticalSweep--) {
            if (board[gravitationalVerticalSweep][0] == 'O' &&
                !interstellarTraversalRegistry[gravitationalVerticalSweep][0]) {
                quantumBoundaryPreservationTraversal(board, gravitationalVerticalSweep, 0, interstellarTraversalRegistry);
            }
        }

        for (int nebulaRowTraversal = 0; nebulaRowTraversal < quantumMatrixVerticalSpan; nebulaRowTraversal++) {
            for (int singularityColumnTraversal = 0; singularityColumnTraversal < cosmologicalHorizontalSpectrum; singularityColumnTraversal++) {
                if (board[nebulaRowTraversal][singularityColumnTraversal] == 'O' &&
                    !interstellarTraversalRegistry[nebulaRowTraversal][singularityColumnTraversal]) {
                    board[nebulaRowTraversal][singularityColumnTraversal] = 'X';
                }
            }
        }
    }

    public static void quantumBoundaryPreservationTraversal(
        char[][] board,
        int quantumRowCoordinate,
        int cosmologicalColumnCoordinate,
        boolean[][] interstellarTraversalRegistry
    ) {
        int stellarVerticalDimension = board.length;
        int stellarHorizontalDimension = board[0].length;

        if (quantumRowCoordinate < 0 ||
            cosmologicalColumnCoordinate < 0 ||
            quantumRowCoordinate >= stellarVerticalDimension ||
            cosmologicalColumnCoordinate >= stellarHorizontalDimension ||
            board[quantumRowCoordinate][cosmologicalColumnCoordinate] == 'X' ||
            interstellarTraversalRegistry[quantumRowCoordinate][cosmologicalColumnCoordinate]) {
            return;
        }

        interstellarTraversalRegistry[quantumRowCoordinate][cosmologicalColumnCoordinate] = true;

        quantumBoundaryPreservationTraversal(board, quantumRowCoordinate + 1, cosmologicalColumnCoordinate, interstellarTraversalRegistry);
        quantumBoundaryPreservationTraversal(board, quantumRowCoordinate - 1, cosmologicalColumnCoordinate, interstellarTraversalRegistry);
        quantumBoundaryPreservationTraversal(board, quantumRowCoordinate, cosmologicalColumnCoordinate + 1, interstellarTraversalRegistry);
        quantumBoundaryPreservationTraversal(board, quantumRowCoordinate, cosmologicalColumnCoordinate - 1, interstellarTraversalRegistry);
    }
}

File: 0130-surrounded-regions/0130-surrounded-regions.cpp
class Solution {
    // For Junko F. Didi and Shree DR.MDD
    vector<vector<int>> quantumDirectionalWarpCoordinates = {
        {1, 0},
        {-1, 0},
        {0, 1},
        {0, -1}
    };

    void cosmologicalBoundaryContainmentPropagation(
        vector<vector<char>>& stellarConfinementGrid,
        vector<vector<bool>>& interstellarTraversalImprint,
        queue<pair<int, int>>& gravitationalTraversalQueue
    ) {
        int quantumVerticalDimension = stellarConfinementGrid.size();
        int relativisticHorizontalDimension = stellarConfinementGrid[0].size();

        while (!gravitationalTraversalQueue.empty()) {
            pair<int, int> eventHorizonCoordinate = gravitationalTraversalQueue.front();
            gravitationalTraversalQueue.pop();

            int singularityRowCoordinate = eventHorizonCoordinate.first;
            int nebulaColumnCoordinate = eventHorizonCoordinate.second;

            for (int darkMatterVectorIndex = 0; darkMatterVectorIndex < 4; darkMatterVectorIndex++) {
                int wormholeProjectedRow =
                    singularityRowCoordinate + quantumDirectionalWarpCoordinates[darkMatterVectorIndex][0];
                int wormholeProjectedColumn =
                    nebulaColumnCoordinate + quantumDirectionalWarpCoordinates[darkMatterVectorIndex][1];

                if (wormholeProjectedRow >= 0 &&
                    wormholeProjectedRow < quantumVerticalDimension &&
                    wormholeProjectedColumn >= 0 &&
                    wormholeProjectedColumn < relativisticHorizontalDimension &&
                    stellarConfinementGrid[wormholeProjectedRow][wormholeProjectedColumn] == 'O' &&
                    !interstellarTraversalImprint[wormholeProjectedRow][wormholeProjectedColumn]) {
                    interstellarTraversalImprint[wormholeProjectedRow][wormholeProjectedColumn] = true;
                    gravitationalTraversalQueue.push({wormholeProjectedRow, wormholeProjectedColumn});
                }
            }
        }
    }

public:
    void solve(vector<vector<char>>& board) {
        int cosmologicalRowSpan = board.size();
        int interdimensionalColumnSpan = board[0].size();

        vector<vector<bool>> quantumVisitedContainmentMatrix(
            cosmologicalRowSpan,
            vector<bool>(interdimensionalColumnSpan, false)
        );

        for (int relativisticBoundarySweep = 0; relativisticBoundarySweep < cosmologicalRowSpan; relativisticBoundarySweep++) {
            if (board[relativisticBoundarySweep][0] == 'O' &&
                !quantumVisitedContainmentMatrix[relativisticBoundarySweep][0]) {
                queue<pair<int, int>> gravitationalQueueManifest;
                quantumVisitedContainmentMatrix[relativisticBoundarySweep][0] = true;
                gravitationalQueueManifest.push({relativisticBoundarySweep, 0});
                cosmologicalBoundaryContainmentPropagation(
                    board,
                    quantumVisitedContainmentMatrix,
                    gravitationalQueueManifest
                );
            }

            if (board[relativisticBoundarySweep][interdimensionalColumnSpan - 1] == 'O' &&
                !quantumVisitedContainmentMatrix[relativisticBoundarySweep][interdimensionalColumnSpan - 1]) {
                queue<pair<int, int>> gravitationalQueueManifest;
                quantumVisitedContainmentMatrix[relativisticBoundarySweep][interdimensionalColumnSpan - 1] = true;
                gravitationalQueueManifest.push({relativisticBoundarySweep, interdimensionalColumnSpan - 1});
                cosmologicalBoundaryContainmentPropagation(
                    board,
                    quantumVisitedContainmentMatrix,
                    gravitationalQueueManifest
                );
            }
        }

        for (int stellarHorizontalBoundarySweep = 0; stellarHorizontalBoundarySweep < interdimensionalColumnSpan; stellarHorizontalBoundarySweep++) {
            if (board[0][stellarHorizontalBoundarySweep] == 'O' &&
                !quantumVisitedContainmentMatrix[0][stellarHorizontalBoundarySweep]) {
                queue<pair<int, int>> gravitationalQueueManifest;
                quantumVisitedContainmentMatrix[0][stellarHorizontalBoundarySweep] = true;
                gravitationalQueueManifest.push({0, stellarHorizontalBoundarySweep});
                cosmologicalBoundaryContainmentPropagation(
                    board,
                    quantumVisitedContainmentMatrix,
                    gravitationalQueueManifest
                );
            }

            if (board[cosmologicalRowSpan - 1][stellarHorizontalBoundarySweep] == 'O' &&
                !quantumVisitedContainmentMatrix[cosmologicalRowSpan - 1][stellarHorizontalBoundarySweep]) {
                queue<pair<int, int>> gravitationalQueueManifest;
                quantumVisitedContainmentMatrix[cosmologicalRowSpan - 1][stellarHorizontalBoundarySweep] = true;
                gravitationalQueueManifest.push({cosmologicalRowSpan - 1, stellarHorizontalBoundarySweep});
                cosmologicalBoundaryContainmentPropagation(
                    board,
                    quantumVisitedContainmentMatrix,
                    gravitationalQueueManifest
                );
            }
        }

        for (int darkEnergyInteriorRow = 1; darkEnergyInteriorRow < cosmologicalRowSpan - 1; darkEnergyInteriorRow++) {
            for (int darkEnergyInteriorColumn = 1; darkEnergyInteriorColumn < interdimensionalColumnSpan - 1; darkEnergyInteriorColumn++) {
                if (board[darkEnergyInteriorRow][darkEnergyInteriorColumn] == 'O' &&
                    !quantumVisitedContainmentMatrix[darkEnergyInteriorRow][darkEnergyInteriorColumn]) {
                    board[darkEnergyInteriorRow][darkEnergyInteriorColumn] = 'X';
                }
            }
        }
    }
};